packages/utils/src/abi_utils.ts

```diff
@@ -1,7 +1,166 @@
 import { AbiDefinition, AbiType, ContractAbi, DataItem, MethodAbi } from 'ethereum-types';
 import * as _ from 'lodash';
 
+import { BigNumber } from './configured_bignumber';
+
+export type EthersParamName = null | string | EthersNestedParamName;
+
+export interface EthersNestedParamName {
+    name: string | null;
+    names: EthersParamName[];
+}
+
+// Note(albrow): This function is unexported in ethers.js. Copying it here for
+// now.
+// Source: https://github.com/ethers-io/ethers.js/blob/884593ab76004a808bf8097e9753fb5f8dcc3067/contracts/interface.js#L30
+function parseEthersParams(params: DataItem[]): { names: EthersParamName[]; types: string[] } {
+    const names: EthersParamName[] = [];
+    const types: string[] = [];
+
+    params.forEach((param: DataItem) => {
+        if (param.components != null) {
+            let suffix = '';
+            const arrayBracket = param.type.indexOf('[');
+            if (arrayBracket >= 0) {
+                suffix = param.type.substring(arrayBracket);
+            }
+
+            const result = parseEthersParams(param.components);
+            names.push({ name: param.name || null, names: result.names });
+            types.push('tuple(' + result.types.join(',') + ')' + suffix);
+        } else {
+            names.push(param.name || null);
+            types.push(param.type);
+        }
+    });
+
+    return {
+        names,
+        types,
+    };
+}
+
+// returns true if x is equal to y and false otherwise. Performs some minimal
+// type conversion and data massaging for x and y, depending on type. name and
+// type should typically be derived from parseEthersParams.
+function isAbiDataEqual(name: EthersParamName, type: string, x: any, y: any): boolean {
+    if (_.isUndefined(x) && _.isUndefined(y)) {
+        return true;
+    } else if (_.isUndefined(x) && !_.isUndefined(y)) {
+        return false;
+    } else if (!_.isUndefined(x) && _.isUndefined(y)) {
+        return false;
+    }
+    if (_.endsWith(type, '[]')) {
+        // For array types, we iterate through the elements and check each one
+        // individually. Strangely, name does not need to be changed in this
+        // case.
+        if (x.length !== y.length) {
+            return false;
+        }
+        const newType = _.trimEnd(type, '[]');
+        for (let i = 0; i < x.length; i++) {
+            if (!isAbiDataEqual(name, newType, x[i], y[i])) {
+                return false;
+            }
+        }
+        return true;
+    }
+    if (_.startsWith(type, 'tuple(')) {
+        if (_.isString(name)) {
+            throw new Error('Internal error: type was tuple but names was a string');
```

Comment:
There are a handful of these sanity checks in this PR. In general, I like to fail fast when something unexpected happens instead of letting things fail later on. Let me know if there is a better/more conventional way to do this.